lib/SIL/SILFunctionType.cpp

```diff
@@ -1059,11 +1059,122 @@ static CanSILFunctionType getSILFunctionType(
     .withIsPseudogeneric(pseudogeneric)
     .withNoEscape(extInfo.isNoEscape());
   
+  // Extract the abstract generic calling convention of the function into a
+  // substituted generic signature for the function type.
+  //
+  // This signature only needs to consider the general calling convention,
+  // so it can reduce away protocol and base class constraints aside from
+  // `AnyObject`. We wanto similar-shaped generic function types to remain
+  // canonically equivalent, like `(T, U) -> ()`, `(T, T) -> ()`,
+  // `(U, T) -> ()` or `(T, T.A) -> ()` when given substitutions that produce
+  // the same function types, so we also introduce a new generic argument for
+  // each position where we see a dependent type, and canonicalize the order in
+  // which we see independent generic arguments.
+  //
+  bool impliedSignature = false;
+  SubstitutionMap substitutions;
+  if (TC.Context.LangOpts.EnableSubstSILFunctionTypesForFunctionValues
+      // We don't currently use substituted function types for generic function
+      // type lowering, though we should for generic methods on classes and
+      // protocols.
+      && !genericSig) {
+    class SubstFunctionTypeCollector {
+    public:
+      TypeConverter &TC;
+      const bool mapReplacementsOutOfContext;
+      
+      SmallVector<GenericTypeParamType *, 4> substGenericParams;
+      SmallVector<Requirement, 4> substRequirements;
+      SmallVector<Type, 4> substReplacements;
+      
+      SubstFunctionTypeCollector(TypeConverter &TC,
+                                 bool mapReplacementsOutOfContext)
+        : TC(TC), mapReplacementsOutOfContext(mapReplacementsOutOfContext) {}
+      SubstFunctionTypeCollector(const SubstFunctionTypeCollector &) = delete;
+
+      // TypeSubstitutionFn
+      Type operator()(SubstitutableType *t) {
+        ArchetypeType *archetype = dyn_cast<ArchetypeType>(t);
+        if (!archetype)
+          archetype = TC.getCurGenericContext()->getGenericEnvironment()
+                                   ->mapTypeIntoContext(t)
+                                   ->castTo<ArchetypeType>();
+        
+        // Replace every dependent type we see with a fresh type variable in
+        // the substituted signature.
```

Comment:
Well, but they are sometimes substituted, right?  On the implementation side of an opaque type, we know that `OpaqueType == Int`.  If we make opaque-type identity part of the ABI, presumably that means that at some point in SIL on the implementation side we'll need to represent function types where the original type is expressed in terms of `OpaqueType` but the concrete type is known to be `Int`.  Feels easier to me to just define this away.

Reply:
The opaque type is never formally substituted; we at best do a convert_function or bitcast in places we know the underlying type. Arnold's work on #28044 should more thoroughly handle the relationship between opaque types and their underlying type as part of type lowering, so the difference would only manifest at boundaries between differently-resilient function bodies.

Reply:
Well, the ABI of a specific function has to be consistent, so it matters if the representation would say that they're ABI-different on different sides of the resilience boundary.  Concretely, if we're going to give `() -> (@out Opaque)` a different ptrauth discriminator from `() -> (@out Int)`, we need to remember that the function signature was actually written as the former, which is what this work is all about; Arnold's work can't just rewrite it on one side of the boundary and in doing so implicitly change the discriminator.

Reply:
Hm, good point. I'll change it to substitute them out.

Reply:
Well, doing that correctly in the face of #28044 might require sinking this more deeply into the type lowering logic than I was hoping to start out with, instead of doing it as a post-pass, since the lowered argument and return types would already have the opaque archetypes substituted out of them. If you don't mind, I can do that in a follow up patch.

Reply:
That's fine, it just needs to happen before we can make ptrauth discriminator tighter.